README
NAME
    Config::Environment - Application Configuration via Environment
    Variables

VERSION
    version 0.000009

SYNOPSIS
        use Config::Environment;

        my $conf = Config::Environment->new('myapp');
        my $conn = $conf->param('db.1.conn' => 'dbi:mysql:dbname=foobar');
        my $user = $conf->param('db.1.user'); # via $ENV{MYAPP_DB_1_USER} or undef
        my $pass = $conf->param('db.1.pass'); # via $ENV{MYAPP_DB_1_PASS} or undef

        or

        my $info = $conf->param('db.1');
        say $info->{conn}; # outputs dbi:mysql:dbname=foobar
        say $info->{user}; # outputs the value of $ENV{MYAPP_DB_1_USER}
        say $info->{pass}; # outputs the value of $ENV{MYAPP_DB_1_PASS}

        likewise ...

        $conf->param('server' => {node => ['10.10.10.02', '10.10.10.03']});

        creates the following environment variables and assignments

        $ENV{MYAPP_SERVER_NODE_1} = '10.10.10.02';
        $ENV{MYAPP_SERVER_NODE_2} = '10.10.10.03';

        ... and the configuration can be retrieved using any of the following

        $conf->param('server');
        $conf->param('server.node');
        $conf->param('server.node.1');
        $conf->param('server.node.2');

        or

        my ($node1, $node2) = $conf->params(qw(server.node.1 server.node.2));

DESCRIPTION
    Config::Environment is an interface for managing application
    configuration using environment variables as a backend. Using
    environment variables as a means of application configuration is a great
    way of controlling which parts of your application configuration gets
    hard-coded and shipped with your codebase (and which parts do not).
    Using environment variables, application configuration can be set at the
    system, user, and/or application levels and easily overridden.

ATTRIBUTES
  autoload
    The autoload attribute contains a boolean value which determines whether
    the global ENV hash will be sourced during instantiation. This attribute
    is set to true by default.

  domain
    The domain attribute contains the environment variable prefix used as
    context to differentiate between other environment variables.

  lifecycle
    The lifecycle attribute contains a boolean value which if true restricts
    any environment variables changes to life of the class instance. This
    attribute is set to false by default.

  mirror
    The mirror attribute contains a boolean value which if true copies any
    configuration assignments to the corresponding environment variables.
    This attribute is set to true by default.

  override
    The override attribute contains a boolean value which determines whether
    parameters corresponding to an existing environment variable can have
    it's value overridden. This attribute is set to true by default.

  stash
    The stash attribute contains a hashref which can be used to store
    arbitrary data which does not undergo parsing and which can be accessed
    using the param method.

METHODS
  load
    The load method expects a hashref which it parses and generates
    environment variables from (whether they exist or not) and registers the
    formatted environment structure. This method is called automatically on
    instantiation using the global ENV hash as an argument. Note! The hash
    can contain nested objects but it's keys should resemble
    capitalized/underscored environment variable names.

        my $hash = {
            APP_MODE => 'development',
            APP_USER => 'vagrant',
            APP_PORT => 9000
        };

        $self->load($hash);

  param
    The param method expects a key which it uses to locate the corresponding
    environment variable in the registered data structure. The key uses
    dot-notation to traverse hierarchical data in the registry. This method
    will return undefined if no element can be found matching the query. The
    method can also be used to set environment variables by passing an
    additional argument as the value in the form of a scalar, arrayref or
    hashref.

        my $item = $self->param($key);
        my $item = $self->param($key => $value);

        # load parsed data from another configuration source, e.g. a config file
        while (my($key, $val) = each(%$configuration) {
            $self->param($key => $val);
        }

  params
    The params method expects a list of keys which are used to locate the
    corresponding environment variables in the registered data structure.
    The keys use dot-notation to traverse hierarchical data in the registry
    and return a list of corresponding values in order specified. This
    method returns a list in list-context, otherwise it returns the first
    element found of the list of queries specified.

        my $item  = $self->params(@list_of_keys);
        my @items = $self->params(@list_of_keys);

    You can also pass a single hash-reference to this method and have it
    traverse the key/value pairs and perform the desired assignments. This
    usage will not return a value.

        $self->params(\%params);

  environment
    The environment method returns a hashref representing all environment
    variables specific to the instantiated object's domain and instance.

        my $environment = $self->environment;

  subdomain
    The subdomain method returns a copy of the class instance with a
    modified domain reference for easier access to nested configuration
    keys.

        my $db  = $self->subdomain('db');
        my $db1 = $db->subdomain('1');

        $db1->param('conn' => $connstring);
        $db1->param('user' => $username);
        $db1->param('pass' => $password);

AUTHOR
    Al Newkirk <anewkirk@ana.io>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Al Newkirk.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2012-12-27 00:00:00 +0000 to present.
==================================================

---------------------------------------------
version 0.000009 at 2013-12-27 05:40:53 +0000
---------------------------------------------

  Change: 36ae0d4f4da6fc5dcec0045957b658a315d01b7a
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-12-27 00:39:15 +0000

    added a stash feature with access via the param method 

  Change: 64e1cce1e5b249a221eeff589f6b5615d349cce2
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-12-26 23:02:32 +0000

    replaced Hash::Merge with Hash::Merge::Simple 

  Change: bf65af1efc1ece8a8dbee10361a8714cdb0022ba
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-24 06:19:27 +0000

    return void from hashref-loading 

---------------------------------------------
version 0.000008 at 2013-10-24 09:30:27 +0000
---------------------------------------------

  Change: 699dd4833d2a3a96e4e3a143977f244dbc310b1e
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-24 05:30:27 +0000

    v0.000008 

  Change: c0dc30b70803c616d30ee91979ae5306af5a3837
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-24 05:30:14 +0000

    load hash-references using the params method 

---------------------------------------------
version 0.000007 at 2013-10-19 02:00:22 +0000
---------------------------------------------

  Change: 30d62591af0a7664e7adc8c214f497ff1712bdbb
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-18 22:00:22 +0000

    v0.000007 

  Change: 9c32164aaacfc2faf4e3303bb4673bb0e6614351
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-18 21:59:31 +0000

    instantiate with a subdomain 

---------------------------------------------
version 0.000006 at 2013-10-06 05:13:14 +0000
---------------------------------------------

  Change: af2985d53c0f2b453dfd405aac495cd64e8ed8ef
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-06 01:13:14 +0000

    v0.000006 

  Change: 9a2e4e717003d843b3e680522d376dddd870c830
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-06 01:13:00 +0000

    update to cleanup internal data structures 

---------------------------------------------
version 0.000005 at 2013-10-06 04:11:55 +0000
---------------------------------------------

  Change: ee360cd400243974c43788c18f10d5cc83cc9d25
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-06 00:11:55 +0000

    v0.000005 

  Change: bf8431b14cf12ddf6a23d68d45fa3d86a3f3f8f3
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-10-06 00:11:29 +0000

    added better handling of parameter assignment and updated the
    documentation 

---------------------------------------------
version 0.000004 at 2013-09-10 05:59:48 +0000
---------------------------------------------

  Change: fb9313e2f53bf682805318a14611e81eeef03cf5
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-10 01:59:48 +0000

    v0.000004 

  Change: dcf5f16cab907db2932009352291898e3bbf4448
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-10 01:59:33 +0000

    added the lifecycle and mirror attributes 

---------------------------------------------
version 0.000003 at 2013-09-08 16:39:57 +0000
---------------------------------------------

  Change: d72cf64644af690cdf8ac8e9f85538e069f46b41
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-08 12:39:57 +0000

    v0.000003 

  Change: 4164a3cabff7fddb19a7a07d828a84fdeac0d181
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-08 12:39:27 +0000

    added the subdomain method and the autoload and override attributes 

---------------------------------------------
version 0.000002 at 2013-09-06 10:18:44 +0000
---------------------------------------------

  Change: 13a5e88d9cf8094c306e09c9cca6fed80e5b39ff
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-06 06:18:44 +0000

    v0.000002 

  Change: 77a03842165f51d2c923da52da8b9d4fc4cf4ead
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-06 06:18:29 +0000

    documentation update and full test coverage 

  Change: 9bb869f6ebc763cce7c6f1552b932321e4b08493
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-06 04:35:41 +0000

    updated documentation 

---------------------------------------------
version 0.000001 at 2013-09-06 08:04:50 +0000
---------------------------------------------

  Change: e4456ff1c4a88451809b5dfadbf6a3aac1731613
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-06 04:04:50 +0000

    v0.000001 

  Change: 8952e857193ee3ecd54b7c36699336143276d03a
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-06 04:04:30 +0000

    initial commit 

================
End of releases.
================


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Application Configuration via Environment Variables",
  "AUTHOR" => "Al Newkirk <anewkirk\@ana.io>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Config-Environment",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Config::Environment",
  "PREREQ_PM" => {
    "Hash::Flatten" => 0,
    "Hash::Merge::Simple" => 0,
    "Moo" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.000009",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Hash::Flatten" => 0,
  "Hash::Merge::Simple" => 0,
  "Moo" => 0,
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Config/Environment.pm
# ABSTRACT: Application Configuration via Environment Variables
package Config::Environment;

use utf8;
use 5.10.0;

use Moo;
use Hash::Flatten ();
use Hash::Merge::Simple ();

our $VERSION = '0.000009'; # VERSION


sub BUILDARGS {
    my ($class, @args) = @_;

    unshift @args, 'domain' if $args[0] && $#args == 0;
    return {@args};
}

sub BUILD {
    my ($self) = @_;

    $self->{domain} = lc $self->{domain};
    if ($self->{domain} =~ s/[^a-zA-Z0-9]+/_/g) {
        my ($dom, $subdom) = split /_/, $self->{domain}, 2;
        $self->{domain}    = $dom;
        $self->{subdomain} = $self->to_sub_key($subdom);
    }

    my $dom = $self->domain;
    $self->{snapshot} = { map {$_ => $ENV{$_}} grep { /^$dom\_/i } keys %ENV };
    return $self->load({%ENV}) if $self->autoload;
}


has autoload => (
    is       => 'ro',
    required => 0,
    default  => 1
);


has domain => (
    is       => 'ro',
    required => 1
);


has lifecycle => (
    is       => 'ro',
    required => 0,
    default  => 0
);


has mirror => (
    is       => 'rw',
    required => 0,
    default  => 1
);


has override => (
    is       => 'rw',
    required => 0,
    default  => 1
);


has stash => (
    is       => 'ro',
    required => 0,
    default  => sub {{}}
);


sub load {
    my ($self, $hash) = @_;
    my $dom = lc $self->domain;
    my $env = { map {$_ => $hash->{$_}} grep { /^$dom\_/i } keys %{$hash} };
    my $reg = $self->{registry} //= {env => {}, map => {}};
    my $map = $reg->{map};

    for my $key (sort keys %{$env}) {
        my $value = delete $env->{$key};

        $key =~ s/_/./g;
        $key =~ s/^$dom\.//gi;

        my $hash = {lc $key => $value};

        if (ref $value) {
            if ('ARRAY' eq ref $value) {
                my $i = 0;
                $value = { map { ++$i => $_ } @{$value} };
            }

            $hash = Hash::Flatten->new->flatten($value);

            for my $refkey (keys %{$hash}) {
                (my $newref = $refkey) =~ s/(\w):(\d+)/"$1.".($2+1)/gpe;
                $hash->{lc "$key.$newref"} = delete $hash->{$refkey};
            }
        }

        $map = Hash::Merge::Simple->merge(
            $map => Hash::Flatten->new->unflatten($hash)
        );

        if ($self->mirror) {
            while (my($key, $val) = each(%{$hash})) {
                $ENV{$self->to_env_key($key)} = $val;
            }
        }
    }

    $reg->{map} = $map;

    return $self;
}


sub param {
    my ($self, $key, $val) = @_;

    return unless defined $key;

    my $dom = $self->domain;

    $key = $self->to_dom_key($key);
    $key =~ s/^$dom(\.)?//;

    if (@_ > 2) {
        my $pairs = Hash::Flatten::flatten({$key => $val});
        while (my($key, $val) = each(%{$pairs})) {
            $key =~ s/(\w):(\d+)/"$1.".($2+1)/gpe;
            $key =~ s/\\//g;
            unless (exists $ENV{$self->to_env_key($key)} && ! $self->override) {
                $self->load({$self->to_env_key($key) => $val});
                $self->{registry}{env}{$key} = $val;
            }
        }
    }

    my $result;

    # env lookup
    if (exists $self->{registry}{env}{$key}) {
        $result = $self->{registry}{env}{$key};
    }

    # env map walk
    if (!$result) {
        my $node  = $self->{registry}{map};
        my @steps = split /\./, $key;
        for (my $i=0; $i<@steps; $i++) {
            my $step = $steps[$i];
            if (exists $node->{$step}) {
                if ($i<@steps && 'HASH' ne ref $node) {
                    undef $node and last;
                }
                $node = $node->{$step};
            }
            else {
                undef $node and last;
            }
        }
        $result = $node;
    }

    # stash walk
    if (!$result) {
        my $key = join '.', grep defined, $self->{subdomain}, $_[1]; #hack
        $key =~ s/\.(\d+)\./".".($1-1)."."/gpe;
        unless ($result = $self->stash->{$key}) {
            my $node  = $self->stash;
            my @steps = split /\./, $key;
            for (my $i=0; $i<@steps; $i++) {
                my $step = $steps[$i];
                if ('ARRAY' eq ref $node) {
                    if ($i<@steps && !defined $node->[$step]) {
                        undef $node and last;
                    }
                    else {
                        $node = $node->[$step];
                    }
                }
                elsif ('HASH' eq ref $node) {
                    if ($i<@steps && !defined $node->{$step}) {
                        undef $node and last;
                    }
                    else {
                        $node = $node->{$step};
                    }
                }
                else {
                    undef $node and last;
                }
            }
            $result = $node;
        }
    }

    return $result;
}


sub params {
    my ($self, @keys) = @_;

    if ($#keys == 0) {
        if ('HASH' eq ref $keys[0]) {
            while (my ($key, $value) = each%{$keys[0]}) {
                $self->param($key, $value);
            }
            return;
        }
    }

    my @vals = map { $self->param($_) } @keys;
    return wantarray ? @vals : $vals[0];
}


sub environment {
    my ($self) = @_;
    my $map = Hash::Flatten->new->flatten($self->{registry}{map});

    for my $key (keys %{$map}) {
        $map->{$self->to_env_key($key)} = delete $map->{$key};
    }

    return $map;
}


sub subdomain {
    my ($self, $key) = @_;
    my $dom  = $self->domain;
    my $copy = ref($self)->new(
        autoload  => 0,
        override  => $self->override,
        lifecycle => $self->lifecycle,
        mirror    => $self->mirror,
        stash     => $self->stash,
        domain    => $dom
    );

    $copy->{subdomain} = $self->to_sub_key($key);
    $copy->{registry}  = $self->{registry};

    return $copy;
}

sub to_dom_key {
    my ($self, $key) = @_;
    my $dom = $self->domain;

    $key =~ s/^$dom//;

    my @prefix = ($dom);
    push @prefix, $self->{subdomain} if defined $self->{subdomain};

    return lc join '.', @prefix, split /_/, $key;
}

sub to_env_key {
    my ($self, $key) = @_;
    my $dom = $self->domain;

    $key =~ s/^$dom//;

    return uc join '_', $dom, split /\./, $key
}

sub to_sub_key {
    my ($self, $key) = @_;
    my $dom = $self->domain;

    ($key = $self->to_dom_key($key)) =~ s/^$dom(\.)?//;

    return $key;
}

sub DESTROY {
    my ($self) = @_;

    if ($self->lifecycle) {
        my $environment = $self->environment;
        my $snapshot    = $self->{snapshot};

        delete $ENV{$_} for grep { ! exists $snapshot->{$_} }
            keys %{$environment};
        $ENV{$_} = $snapshot->{$_} for keys %{$snapshot};
    }
}

1;

__END__

=pod

=head1 NAME

Config::Environment - Application Configuration via Environment Variables

=head1 VERSION

version 0.000009

=head1 SYNOPSIS

    use Config::Environment;

    my $conf = Config::Environment->new('myapp');
    my $conn = $conf->param('db.1.conn' => 'dbi:mysql:dbname=foobar');
    my $user = $conf->param('db.1.user'); # via $ENV{MYAPP_DB_1_USER} or undef
    my $pass = $conf->param('db.1.pass'); # via $ENV{MYAPP_DB_1_PASS} or undef

    or

    my $info = $conf->param('db.1');
    say $info->{conn}; # outputs dbi:mysql:dbname=foobar
    say $info->{user}; # outputs the value of $ENV{MYAPP_DB_1_USER}
    say $info->{pass}; # outputs the value of $ENV{MYAPP_DB_1_PASS}

    likewise ...

    $conf->param('server' => {node => ['10.10.10.02', '10.10.10.03']});

    creates the following environment variables and assignments

    $ENV{MYAPP_SERVER_NODE_1} = '10.10.10.02';
    $ENV{MYAPP_SERVER_NODE_2} = '10.10.10.03';

    ... and the configuration can be retrieved using any of the following

    $conf->param('server');
    $conf->param('server.node');
    $conf->param('server.node.1');
    $conf->param('server.node.2');

    or

    my ($node1, $node2) = $conf->params(qw(server.node.1 server.node.2));

=head1 DESCRIPTION

Config::Environment is an interface for managing application configuration using
environment variables as a backend. Using environment variables as a means of
application configuration is a great way of controlling which parts of your
application configuration gets hard-coded and shipped with your codebase (and
which parts do not). Using environment variables, application configuration can
be set at the system, user, and/or application levels and easily overridden.

=head1 ATTRIBUTES

=head2 autoload

The autoload attribute contains a boolean value which determines whether
the global ENV hash will be sourced during instantiation. This attribute is
set to true by default.

=head2 domain

The domain attribute contains the environment variable prefix used as context
to differentiate between other environment variables.

=head2 lifecycle

The lifecycle attribute contains a boolean value which if true restricts any
environment variables changes to life of the class instance. This attribute
is set to false by default.

=head2 mirror

The mirror attribute contains a boolean value which if true copies any
configuration assignments to the corresponding environment variables. This
attribute is set to true by default.

=head2 override

The override attribute contains a boolean value which determines whether
parameters corresponding to an existing environment variable can have it's
value overridden. This attribute is set to true by default.

=head2 stash

The stash attribute contains a hashref which can be used to store arbitrary data
which does not undergo parsing and which can be accessed using the param method.

=head1 METHODS

=head2 load

The load method expects a hashref which it parses and generates environment
variables from (whether they exist or not) and registers the formatted
environment structure. This method is called automatically on instantiation
using the global ENV hash as an argument. Note! The hash can contain nested
objects but it's keys should resemble capitalized/underscored environment
variable names.

    my $hash = {
        APP_MODE => 'development',
        APP_USER => 'vagrant',
        APP_PORT => 9000
    };

    $self->load($hash);

=head2 param

The param method expects a key which it uses to locate the corresponding
environment variable in the registered data structure. The key uses dot-notation
to traverse hierarchical data in the registry. This method will return undefined
if no element can be found matching the query. The method can also be used to
set environment variables by passing an additional argument as the value in the
form of a scalar, arrayref or hashref.

    my $item = $self->param($key);
    my $item = $self->param($key => $value);

    # load parsed data from another configuration source, e.g. a config file
    while (my($key, $val) = each(%$configuration) {
        $self->param($key => $val);
    }

=head2 params

The params method expects a list of keys which are used to locate the
corresponding environment variables in the registered data structure. The keys
use dot-notation to traverse hierarchical data in the registry and return a list
of corresponding values in order specified. This method returns a list in
list-context, otherwise it returns the first element found of the list of
queries specified.

    my $item  = $self->params(@list_of_keys);
    my @items = $self->params(@list_of_keys);

You can also pass a single hash-reference to this method and have it traverse
the key/value pairs and perform the desired assignments. This usage will not
return a value.

    $self->params(\%params);

=head2 environment

The environment method returns a hashref representing all environment variables
specific to the instantiated object's domain and instance.

    my $environment = $self->environment;

=head2 subdomain

The subdomain method returns a copy of the class instance with a modified domain
reference for easier access to nested configuration keys.

    my $db  = $self->subdomain('db');
    my $db1 = $db->subdomain('1');

    $db1->param('conn' => $connstring);
    $db1->param('user' => $username);
    $db1->param('pass' => $password);

=head1 AUTHOR

Al Newkirk <anewkirk@ana.io>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Al Newkirk.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
